### DISCUSSION TITLE
Looking for a high-performance actor library.

### DISCUSSION THREAD

**Discussion by zyansheep** [slawlor/ractor#92] [General]
Hello! I'm looking for a high-performance actor library for a low-latency p2p networking application and I'm curious what kind of performance characteristics this library has and how it compares to other actor libraries out there.

Some questions:
 - Does this library limit allocations wherever possible?
 - Does this library limit context switches when doing a blocking call from one actor to another?
 - Can actors interface well with high-throughput I/O?
 - Can the input message buffer for any given actor be customized for certain latency / thoughput requirements?
 - What kind of overhead is there for a given actor? If someone attempts to DOS a p2p node running ractor with millions of incoming connections, and each connection spawns at least one actor to handle packet codec, cryptography, and metrics, should I be worried about potential OOMs?
 - Is there any plans for ractor to be runtime-agnostic? (currently using futures and async-std for my project)
 
 I may have more questions, but that's all I can think of now. Overall cool library tho :)

**Comment by slawlor** [comment]
I think my answer in #94 might answer a lot of this, so please see there for some information.

> Does this library limit allocations wherever possible?

We try to stay away from `Clone` as much as possible, which is why everything must be `Send` to move it around across threads when necessary.

> Does this library limit context switches when doing a blocking call from one actor to another?

This is left to `tokio`, as we just await an underlying channel's reply. I think it's quite likely you'll end up on another thread, but it'll get polled like a normal future.

> Can actors interface well with high-throughput I/O?

Depending on how you want to handle said I/O, you can within an actor do a `spawn_blocking` to get an CPU bound specific thread for high CPU-based work. Actor's themselves shouldn't block a thread ideally, as they're all scheduled on the tokio high-performance threadpool. However async I/O is fine to do in an actor, and you can make it as performant as you want, either looping the actor, or blocking the message handler and doing all your I/O. That's up to you in your design.

> Can the input message buffer for any given actor be customized for certain latency / thoughput requirements?

No, but a [factory](https://docs.rs/ractor/0.8.4/ractor/factory/index.html) can handle this for you adding latency + queueing management along with routing control.

> ... should I be worried about potential OOMs?

You should always be worried about OOMs :) However we're really just a tokio task with some data (see [`ActorProperties`](https://github.com/slawlor/ractor/blob/main/ractor/src/actor/actor_cell/actor_properties.rs#LL22C1-L36C2)) describing the actor. That being said, there's only ever 1 instance of `ActorProperties` for an actor, all `ActorCell` or `ActorRef`'s are `Arc`s of that struct instance which is thread-safe. Additionally each actor holds a state, which can be as large or small as you define, so that part's up to you.

That being said, except for the properties, you can spawn as many actors as tokio tasks without much worry for memory management.

> Is there any plans for ractor to be runtime-agnostic? (currently using futures and async-std for my project)

So yes, there's partial support for `async-std` but I haven't gotten it fully working yet. There's some problems with `JoinHandle` between `tokio` & `async-std`'s version, along with the polling `select!` implementation. You are welcome to try and round out the primitives in the `ractor::concurrency` module. It's half-baked for now...

That being said, we do hard rely on Tokio's channels implementation (which is agnostic of the runtime, so is `async-std` compatible). You can import `tokio` with the `sync` feature without being in a `tokio` runtime :)

**Reply by zyansheep** [reply]
Thanks for the response! Looks like there is still much work to be done :eyes: 

Ideally, high-throughput I/O would use something like `mio` on the backend and resume an actor's handler future immediately when the kernel polls the program notifying that there is new data to read. (Or through a mechanism like io_uring). Seems like it should be easy enough to just await a TcpStream in an actor's async message handler and then loop :)

Thanks for making this cool library btw, I can't wait to see how it evolves into the future!

Btw, have you considered creating a discord / matrix server for more real-time communication?
